File: zig/src/p_0016.zig
const std = @import("std");

// Function to compute the sum of the digits of 2^1000
fn euler_0016(comptime exponent: usize) usize {
    // Initialize an array to store the digits of 2^exponent
    var digits: [exponent]u8 = std.mem.zeroes([exponent]u8);
    digits[0] = 1; // Start with 2^0 = 1

    var carry: u8 = 0;
    var length: usize = 1; // Current number of digits

    // Multiply by 2 repeatedly to compute 2^exponent
    for (0..exponent) |_| {
        carry = 0;
        for (0..length) |i| {
            const product = digits[i] * 2 + carry;
            digits[i] = @as(u8, product % 10);
            carry = @as(u8, product / 10);
        }

        // If there's a carry after the last digit, increase the length
        while (carry > 0) {
            digits[length] = @as(u8, carry % 10);
            carry /= 10;
            length += 1;
        }
    }

    // Sum the digits
    var sum: usize = 0;
    for (digits[0..length]) |digit| {
        sum += digit;
    }

    return sum;
}

test "euler16" {
    try std.testing.expectEqual(26, euler_0016(15));
}

pub fn main() !void {
    std.debug.print("Problem 0016: The sum of the digits of 2^1000 is: {d}\n", .{euler_0016(1000)});
}

File: cpp/src/p_0016.cc
#include "euler.h"

// Compute the sum of the digits of 2^exponent
int euler_0016(int exponent) {
    // Initialize a vector to store the digits of 2^exponent
    std::vector<int> digits(exponent, 0);
    digits[0] = 1;  // Start with 2^0 = 1

    int carry{0};
    int length{1};  // Current number of digits

    // Multiply by 2 repeatedly to compute 2^exponent
    for (int i{0}; i < exponent; ++i) {
        carry = 0;
        for (int j{0}; j < length; ++j) {
            int product{digits[j] * 2 + carry};
            digits[j] = product % 10;
            carry = product / 10;
        }

        // If there's a carry after the last digit, increase the length
        while (carry > 0) {
            digits[length++] = carry % 10;
            carry /= 10;
        }
    }

    return std::accumulate(digits.begin(), digits.begin() + length, 0);
}
